consumer/file.go

```diff
@@ -280,8 +287,186 @@ func (cons *File) Consume(workers *sync.WaitGroup) {
+	printFileOpenError bool
+}
+
+func (source *sourceFile) Configure(conf core.PluginConfigReader) {
+	source.realFileName = source.getRealFileName()
+	source.printFileOpenError = true
+}
+
+func (source *sourceFile) getRealFileName() string {
+	baseFileName, err := filepath.EvalSymlinks(source.fileName)
+	if err != nil {
+		baseFileName = source.fileName
+	}
+
+	baseFileName, err = filepath.Abs(baseFileName)
+	if err != nil {
+		baseFileName = source.fileName
+	}
+
+	return baseFileName
+}
+
+func newSourceFile(conf core.PluginConfigReader) (sourceFile, error) {
+	source := sourceFile{}
+	err := conf.Configure(&source)
+	if err != nil {
+		return source, err
+	}
+
+	return source, nil
+}
+
+// -- seeker --
+
+type seeker struct {
+	seek     int
+	onRotate int
+	offset   int64
+}
+
+func newSeeker(conf core.PluginConfigReader) seeker {
+	switch strings.ToLower(conf.GetString("DefaultOffset", fileOffsetEnd)) {
+	default:
+		fallthrough
+	case fileOffsetEnd:
+		return seeker{
+			seek:     2,
+			onRotate: 1,
+			offset:   0,
+		}
+
+	case fileOffsetStart:
+		return seeker{
+			seek:     1,
+			onRotate: 1,
+			offset:   0,
+		}
+	}
+}
+
+// -- watcher --
+
+type watcher struct {
+	logger logrus.FieldLogger
+	source *sourceFile
+	read   func()
+
+	done  chan int
+	retry chan int
+}
+
+func newWatcher(logger logrus.FieldLogger, source *sourceFile, readFunction func()) *watcher {
+	watcher := watcher{}
+
+	watcher.logger = logger
+	watcher.source = source
+	watcher.read = readFunction
+
+	return &watcher
+}
+
+func (w *watcher) close(watcher *fsnotify.Watcher) {
+	err := watcher.Close()
+	if err != nil {
+		w.logger.Fatal(err)
+	}
+}
+
+func (w *watcher) Watch(buffer *tio.BufferedReader, sendFunction func(data []byte)) {
+	// init
+	w.done = make(chan int)
+	w.retry = make(chan int)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		w.logger.Fatal(err)
+	}
+	defer w.close(watcher)
+
+	// watcher init loop
+	go func() {
+		for {
+			if _, err := os.Stat(w.source.realFileName); os.IsNotExist(err) {
+				w.logger.WithField("file", w.source.realFileName).
+					Warning("watched file not exists. retry in 3sec ..")
+				time.Sleep(3 * time.Second)
+				continue // retry
+			}
+
+			// read current file state before watching
+			w.read()
+
+			err = watcher.Add(w.source.realFileName)
+			if err != nil {
+				w.logger.Error("error during adding watcher: ", err)
+				time.Sleep(3 * time.Second)
+				continue // retry
+			}
+
+			// start watch loop to handle file events
+			go w.watchLoop(watcher)
+
+			// handle channel operations 'retry' and 'done'
+			select {
+			case <-w.retry:
+				watcher.Remove(w.source.realFileName)
+				//todo: reset reader?
+
+				continue
+			case <-w.done:
+				return
+			}
+		}
+	}()
+
+	// busy loop for shutdown
+	for w.source.state != fileStateDone {
+		time.Sleep(time.Second)
+	}
+
+	w.logger.Debug("shutdown file watcher ..")
+	close(w.done)
+}
+
+func (w *watcher) watchLoop(watcher *fsnotify.Watcher) {
+	for {
```

Comment:
if this is a select, stopping at done, why the select in the calling function?
This can be moved up.